File: typescript/src/04-extract-and-override-call/order-confirmation-service.ts
import {sendEmail} from "./email-transport.js";
import {orderSubtotalCents, type Order} from "../domain.js";

export interface Confirmation {
  orderId: string;
  emailedTo: string;
  totalCents: number;
}

/**
 * Sends the customer their order confirmation and returns a record of it.
 *
 * The logic we want to test: the confirmation goes to the customer's email, the
 * subject names the order, and the recorded total matches the subtotal.
 *
 * The obstacle: `confirm()` calls the module-level `sendEmail` function
 * *directly*. There is no object to swap out and no constructor parameter to
 * intercept — the dependency is baked into the body of the method as a bare
 * call. The test below therefore hits the real transport, which throws.
 */
export class OrderConfirmationService {

  async confirm(order: Order): Promise<Confirmation> {
    const total = orderSubtotalCents(order);
    const subject = `Your order ${order.id} is confirmed`;
    const body = [
      `Hi ${order.customer.name},`,
      ``,
      `Thanks for your order ${order.id}.`,
      `Total charged: ${(total / 100).toFixed(2)} USD.`,
    ].join("\n");
    let customerEmail = order.customer.email;
    await this.dispatchEmail(customerEmail, subject, body);

    return {
      orderId: order.id,
      emailedTo: customerEmail,
      totalCents: total,
    };
  }

  protected async dispatchEmail(recipientEmail: string, subject: string, body: string) {
    await sendEmail(recipientEmail, subject, body);
  }
}

File: java/src/main/java/com/qmates/kata/extractandoverridecall/OrderConfirmationService.java
package com.qmates.kata.extractandoverridecall;

import com.qmates.kata.domain.Order;
import com.qmates.kata.domain.Orders;

/**
 * Sends the customer their order confirmation and returns a record of it.
 *
 * <p>The logic we want to test: the confirmation goes to the customer's email,
 * the subject names the order, and the recorded total matches the subtotal.
 *
 * <p>The obstacle: {@link #confirm(Order)} calls the static method
 * {@code EmailTransport.sendEmail(...)} <em>directly</em>. There is no object to
 * swap out and no constructor parameter to intercept — the dependency is baked
 * into the body of the method as a bare static call. A test therefore hits the
 * real transport, which throws.
 */
public class OrderConfirmationService {

  public Confirmation confirm(Order order) {
    long total = Orders.subtotalCents(order);
    String subject = "Your order " + order.id() + " is confirmed";
    String body =
        String.join(
            "\n",
            "Hi " + order.customer().name() + ",",
            "",
            "Thanks for your order " + order.id() + ".",
            "Total charged: " + String.format("%.2f", total / 100.0) + " USD.");
    String recipientEmailAddress = order.customer().email();

    sendEmail(recipientEmailAddress, subject, body);

    return new Confirmation(order.id(), recipientEmailAddress, total);
  }

  protected void sendEmail(String recipientEmailAddress, String subject, String body) {
    EmailTransport.sendEmail(recipientEmailAddress, subject, body);
  }

}
